c++ error with conversion to integer

Using latest anaconda python, latest boost, latest pythran and
latest clang, the following code snippet causes a compile error:

import numpy


# pythran export bug(float[])
def bug(x):

        ints=numpy.zeros(5,'int')
        ints[0]=int(numpy.floor(x))

        return ints

#print bug(2.1)

The error is as follows:

%pythran bug.py

CRITICAL Leave the vessel! Women and children first!
C++ compiler failed to compile translated code.
E: Compile error!

******** Command line was: ********
'clang' '/tmp/tmpXBA5cx.cpp' '-I/opt/anaconda/include/python2.7' '-I/opt/anaconda/lib/python2.7/site-packages/numpy/core/include/numpy' '-I/opt/anaconda/lib/python2.7/site-packages/.' '-I/opt/anaconda/lib/python2.7/site-packages/pythran' '-I/opt/anaconda/lib/python2.7/site-packages/pythran/pythonic/patch' '-DENABLE_PYTHON_MODULE' '-DUSE_GMP' '-O2' '-g' '-std=c++11' '-O2' '-shared' '-o' 'bug.so' '-L/opt/anaconda/lib/python2.7/config' '-lpthread' '-ldl' '-lutil' '-lpython2.7' '-lboost_python' '-lgmp' '-lgmpxx' '-fPIC' '-ltcmalloc_minimal'

******** Output :  ********

In file included from /tmp/tmpXBA5cx.cpp:11:
/opt/anaconda/lib/python2.7/site-packages/pythran/pythonic/__builtin__/int_.hpp:13:24: error: no viable conversion from 'pythonic::types::numpy_expr<pythonic::numpy::proxy::floor, pythonic::types::ndarray<double, 1> >' to 'long'
                return t;
                       ^
/opt/anaconda/lib/python2.7/site-packages/pythran/pythonic/__builtin__/int_.hpp:25:38: note: in instantiation of function template specialization 'pythonic::__builtin__::int_<pythonic::types::numpy_expr<pythonic::numpy::proxy::floor, pythonic::types::ndarray<double, 1> > >' requested here
        PROXY(pythonic::__builtin__, int_);
                                     ^
/opt/anaconda/lib/python2.7/site-packages/pythran/pythonic/utils/proxy.hpp:21:35: note: expanded from macro 'PROXY'
#define PROXY(ns,f) FPROXY(ns, f, f)
                                  ^
/opt/anaconda/lib/python2.7/site-packages/pythran/pythonic/utils/proxy.hpp:15:28: note: expanded from macro 'FPROXY'
                return ns::f(std::forward<Types>(types)...); \
                           ^
/tmp/tmpXBA5cx.cpp:66:25: note: in instantiation of function template specialization 'pythonic::__builtin__::proxy::int_::operator()<pythonic::types::numpy_expr<pythonic::numpy::proxy::floor, pythonic::types::ndarray<double, 1> > >' requested here
    std::get<0>(ints) = pythonic::__builtin__::proxy::int_{}(pythonic::numpy::proxy::floor{}(x));
                        ^
/tmp/tmpXBA5cx.cpp:77:30: note: in instantiation of function template specialization '__pythran_bug::bug::operator()<pythonic::types::ndarray<double, 1> >' requested here
  return __pythran_bug::bug()(a0);
                             ^
1 error generated.

--- a/pythran/pythonic/__builtin__/str/join.hpp
+++ b/pythran/pythonic/__builtin__/str/join.hpp
@@ -11,7 +11,7 @@ namespace pythonic { namespace __builtin__  {
         template <class S>
         types::str
         join(S const &s, types::str const& iterable) {
-          long ssize = std::distance(std::begin(s), std::end(s)) - std::is_same<S, types::str>::value ? 0 : 1;
+          long ssize = std::distance(std::begin(s), std::end(s)) - (std::is_same<S, types::str>::value ? 0 : 1);
           /* first iterate over iterable to gather sizes */
           size_t n = ssize * (iterable.size() - 1) + iterable.size();
 

--- a/pythran/pythonic/numpy/float32.hpp
+++ b/pythran/pythonic/numpy/float32.hpp
@@ -2,13 +2,23 @@
 #define PYTHONIC_NUMPY_FLOAT32_HPP
 
 #include "pythonic/utils/proxy.hpp"
+#include "pythonic/utils/meta.hpp"
+#include "pythonic/utils/numpy_traits.hpp"
+#include "pythonic/types/numpy_op_helper.hpp"
 
 namespace pythonic {
 
     namespace numpy {
 
+    namespace details {
+
         float float32() {return float();}
-        PROXY(pythonic::numpy, float32);
+        template<class V> float float32(V v) {return v;}
+    }
+
+#define NUMPY_NARY_FUNC_NAME float32
+#define NUMPY_NARY_FUNC_SYM details::float32
+#include "pythonic/types/numpy_nary_expr.hpp"
 
     }
 

--- a/pythran/pythonic/numpy/float64.hpp
+++ b/pythran/pythonic/numpy/float64.hpp
@@ -2,12 +2,22 @@
 #define PYTHONIC_NUMPY_FLOAT64_HPP
 
 #include "pythonic/utils/proxy.hpp"
+#include "pythonic/utils/meta.hpp"
+#include "pythonic/utils/numpy_traits.hpp"
+#include "pythonic/types/numpy_op_helper.hpp"
 
 namespace pythonic {
 
     namespace numpy {
+    namespace details {
+
         double float64() {return double();}
-        PROXY(pythonic::numpy, float64);
+        template<class V> double float64(V v) {return v;}
+    }
+
+#define NUMPY_NARY_FUNC_NAME float64
+#define NUMPY_NARY_FUNC_SYM details::float64
+#include "pythonic/types/numpy_nary_expr.hpp"
 
     }
 

--- a/pythran/pythonic/numpy/int16.hpp
+++ b/pythran/pythonic/numpy/int16.hpp
@@ -2,16 +2,25 @@
 #define PYTHONIC_NUMPY_INT16_HPP
 
 #include "pythonic/utils/proxy.hpp"
+#include "pythonic/utils/meta.hpp"
+#include "pythonic/utils/numpy_traits.hpp"
+#include "pythonic/types/numpy_op_helper.hpp"
 
 namespace pythonic {
 
     namespace numpy {
 
-        int16_t int16() {return int16_t();}
-        PROXY(pythonic::numpy, int16);
+    namespace details {
 
+        int16_t int16() {return int16_t();}
+        template<class V> int16_t int16(V v) {return v;}
     }
 
+#define NUMPY_NARY_FUNC_NAME int16
+#define NUMPY_NARY_FUNC_SYM details::int16
+#include "pythonic/types/numpy_nary_expr.hpp"
+
+    }
 }
 
 #endif

--- a/pythran/pythonic/numpy/int32.hpp
+++ b/pythran/pythonic/numpy/int32.hpp
@@ -2,13 +2,23 @@
 #define PYTHONIC_NUMPY_INT32_HPP
 
 #include "pythonic/utils/proxy.hpp"
+#include "pythonic/utils/meta.hpp"
+#include "pythonic/utils/numpy_traits.hpp"
+#include "pythonic/types/numpy_op_helper.hpp"
 
 namespace pythonic {
 
     namespace numpy {
 
+    namespace details {
+
         int32_t int32() {return int32_t();}
-        PROXY(pythonic::numpy, int32);
+        template<class V> int32_t int32(V v) {return v;}
+    }
+
+#define NUMPY_NARY_FUNC_NAME int32
+#define NUMPY_NARY_FUNC_SYM details::int32
+#include "pythonic/types/numpy_nary_expr.hpp"
 
     }
 

--- a/pythran/pythonic/numpy/int64.hpp
+++ b/pythran/pythonic/numpy/int64.hpp
@@ -2,13 +2,23 @@
 #define PYTHONIC_NUMPY_INT64_HPP
 
 #include "pythonic/utils/proxy.hpp"
+#include "pythonic/utils/meta.hpp"
+#include "pythonic/utils/numpy_traits.hpp"
+#include "pythonic/types/numpy_op_helper.hpp"
 
 namespace pythonic {
 
     namespace numpy {
 
+    namespace details {
+
         int64_t int64() {return int64_t();}
-        PROXY(pythonic::numpy, int64);
+        template<class V> int64_t int64(V v) {return v;}
+    }
+
+#define NUMPY_NARY_FUNC_NAME int64
+#define NUMPY_NARY_FUNC_SYM details::int64
+#include "pythonic/types/numpy_nary_expr.hpp"
 
     }
 

--- a/pythran/pythonic/numpy/int8.hpp
+++ b/pythran/pythonic/numpy/int8.hpp
@@ -2,17 +2,26 @@
 #define PYTHONIC_NUMPY_INT8_HPP
 
 #include "pythonic/utils/proxy.hpp"
+#include "pythonic/utils/meta.hpp"
+#include "pythonic/utils/numpy_traits.hpp"
+#include "pythonic/types/numpy_op_helper.hpp"
 
 namespace pythonic {
 
     namespace numpy {
 
-        uint8_t int8() {return int8_t();}
-        PROXY(pythonic::numpy, int8);
+    namespace details {
+
+        int8_t int8() {return int8_t();}
+        template<class V> int8_t int8(V v) {return v;}
+    }
+
+#define NUMPY_NARY_FUNC_NAME int8
+#define NUMPY_NARY_FUNC_SYM details::int8
+#include "pythonic/types/numpy_nary_expr.hpp"
 
     }
 
 }
 
 #endif
-

--- a/pythran/pythonic/numpy/uint16.hpp
+++ b/pythran/pythonic/numpy/uint16.hpp
@@ -2,13 +2,24 @@
 #define PYTHONIC_NUMPY_UINT16_HPP
 
 #include "pythonic/utils/proxy.hpp"
+#include "pythonic/utils/meta.hpp"
+#include "pythonic/utils/numpy_traits.hpp"
+#include "pythonic/types/numpy_op_helper.hpp"
 
 namespace pythonic {
 
     namespace numpy {
 
+    namespace details {
+
         uint16_t uint16() {return uint16_t();}
-        PROXY(pythonic::numpy, uint16);
+        template<class V> uint16_t uint16(V v) {return v;}
+    }
+
+#define NUMPY_NARY_FUNC_NAME uint16
+#define NUMPY_NARY_FUNC_SYM details::uint16
+#include "pythonic/types/numpy_nary_expr.hpp"
+
 
     }
 

--- a/pythran/pythonic/numpy/uint32.hpp
+++ b/pythran/pythonic/numpy/uint32.hpp
@@ -2,13 +2,23 @@
 #define PYTHONIC_NUMPY_UINT32_HPP
 
 #include "pythonic/utils/proxy.hpp"
+#include "pythonic/utils/meta.hpp"
+#include "pythonic/utils/numpy_traits.hpp"
+#include "pythonic/types/numpy_op_helper.hpp"
 
 namespace pythonic {
 
     namespace numpy {
 
+     namespace details {
+
         uint32_t uint32() {return uint32_t();}
-        PROXY(pythonic::numpy, uint32);
+        template<class V> uint32_t uint32(V v) {return v;}
+    }
+
+#define NUMPY_NARY_FUNC_NAME uint32
+#define NUMPY_NARY_FUNC_SYM details::uint32
+#include "pythonic/types/numpy_nary_expr.hpp"
 
     }
 

--- a/pythran/pythonic/numpy/uint64.hpp
+++ b/pythran/pythonic/numpy/uint64.hpp
@@ -2,13 +2,24 @@
 #define PYTHONIC_NUMPY_UINT64_HPP
 
 #include "pythonic/utils/proxy.hpp"
+#include "pythonic/utils/meta.hpp"
+#include "pythonic/utils/numpy_traits.hpp"
+#include "pythonic/types/numpy_op_helper.hpp"
 
 namespace pythonic {
 
     namespace numpy {
 
+     namespace details {
+
         uint64_t uint64() {return uint64_t();}
-        PROXY(pythonic::numpy, uint64);
+        template<class V> uint64_t uint64(V v) {return v;}
+    }
+
+#define NUMPY_NARY_FUNC_NAME uint64
+#define NUMPY_NARY_FUNC_SYM details::uint64
+#include "pythonic/types/numpy_nary_expr.hpp"
+
 
     }
 

--- a/pythran/pythonic/numpy/uint8.hpp
+++ b/pythran/pythonic/numpy/uint8.hpp
@@ -2,13 +2,24 @@
 #define PYTHONIC_NUMPY_UINT8_HPP
 
 #include "pythonic/utils/proxy.hpp"
+#include "pythonic/utils/meta.hpp"
+#include "pythonic/utils/numpy_traits.hpp"
+#include "pythonic/types/numpy_op_helper.hpp"
 
 namespace pythonic {
 
     namespace numpy {
 
+    namespace details {
+
         uint8_t uint8() {return uint8_t();}
-        PROXY(pythonic::numpy, uint8);
+        template<class V> uint8_t uint8(V v) {return v;}
+    }
+
+#define NUMPY_NARY_FUNC_NAME uint8
+#define NUMPY_NARY_FUNC_SYM details::uint8
+#include "pythonic/types/numpy_nary_expr.hpp"
+
 
     }
 

--- a/pythran/pythonic/types/numpy_op_helper.hpp
+++ b/pythran/pythonic/types/numpy_op_helper.hpp
@@ -32,6 +32,8 @@ template<class... Types>
 struct valid_numexpr_parameters {
   static constexpr bool value = any_numexpr_arg<Types...>::value and all_valid_arg<Types...>::value;
 };
+template<>
+struct valid_numexpr_parameters<> : std::integral_constant<bool, false> {};
 
 
 template<class T0, class T1,

--- a/pythran/pythonic/types/vectorizable_type.hpp
+++ b/pythran/pythonic/types/vectorizable_type.hpp
@@ -26,6 +26,16 @@ namespace numpy {
     struct round_;
     struct signbit;
     struct where;
+    struct int8;
+    struct int16;
+    struct int32;
+    struct int64;
+    struct uint8;
+    struct uint16;
+    struct uint32;
+    struct uint64;
+    struct float32;
+    struct float64;
   }
 }
 
@@ -59,7 +69,20 @@ template<class O>
                               // conditional processing doesn't permit SIMD
                               not std::is_same<O, numpy::proxy::nan_to_num>::value and
                               not std::is_same<O, numpy::proxy::asarray_chkfinite>::value and
-                              not std::is_same<O, numpy::proxy::where>::value
+                              not std::is_same<O, numpy::proxy::where>::value and
+                              // transtyping
+                              not std::is_same<O, numpy::proxy::int8>::value and
+                              not std::is_same<O, numpy::proxy::int16>::value and
+                              not std::is_same<O, numpy::proxy::int32>::value and
+                              not std::is_same<O, numpy::proxy::int64>::value and
+                              not std::is_same<O, numpy::proxy::uint8>::value and
+                              not std::is_same<O, numpy::proxy::uint16>::value and
+                              not std::is_same<O, numpy::proxy::uint32>::value and
+                              not std::is_same<O, numpy::proxy::uint64>::value and
+                              not std::is_same<O, numpy::proxy::float32>::value and
+                              not std::is_same<O, numpy::proxy::float64>::value and
+                              //
+                              true
                               ;
   };
 

--- a/pythran/pythonic/utils/meta.hpp
+++ b/pythran/pythonic/utils/meta.hpp
@@ -6,6 +6,8 @@ namespace pythonic {
 namespace utils {
 
 template<bool... V> struct all_of;
+template<> struct all_of<> : std::integral_constant<bool, true> {};
+
 template<bool V0, bool... V> struct all_of<V0, V...> {
   static constexpr bool value = V0 and all_of<V...>::value;
 };
@@ -14,6 +16,7 @@ template<bool V0> struct all_of<V0> {
 };
 
 template<bool... V> struct any_of;
+template<> struct any_of<> : std::integral_constant<bool, false> {};
 template<bool V0, bool... V> struct any_of<V0, V...> {
   static constexpr bool value = V0 or any_of<V...>::value;
 };

--- a/pythran/tests/test_env.py
+++ b/pythran/tests/test_env.py
@@ -1,7 +1,8 @@
 """ Base file for all Pythran tests. """
 
 from imp import load_dynamic
-from numpy import int32
+from numpy import int8, int16, int32, int64, uint8, uint16, uint32, uint64
+from numpy import float32, float64
 from numpy import ndarray, isnan, isinf, isneginf, complex128, complex64, bool_
 import copy
 import glob
@@ -35,11 +36,13 @@ class TestEnv(unittest.TestCase):
         print "Type of Pythran res : ", type(res)
         print "Type of Python ref : ", type(ref)
         type_matching = (((list, tuple), (list, tuple)),
-                         (float, (int, float)),
+                         (float, (int, float, float32, float64)),
+                         (float32, (int, float, float32)),
+                         (float64, (int, float, float32, float64)),
                          (long, (int, long)),
                          (bool, (bool, bool_)),
                          # FIXME combiner for boolean doesn't work
-                         (int, (int, bool, bool_, int32)))
+                         (int, (int, bool, bool_, int8, int16, int32, int64, uint8, uint16, uint32, uint64)))
         if isinstance(ref, ndarray):
             # res can be an ndarray of dim 0 because of isneginf call
             if ref.ndim == 0 and (not isinstance(res, ndarray)

--- a/pythran/tests/test_numpy_ufunc_unary.py
+++ b/pythran/tests/test_numpy_ufunc_unary.py
@@ -14,14 +14,15 @@ unary_ufunc = (
         'ceil', 'conj', 'conjugate', 'cos', 'cosh',
         'deg2rad', 'degrees',
         'exp', 'expm1',
-        'fabs', 'floor',
-        'isinf', 'isneginf', 'isposinf', 'isnan', 'invert', 'isfinite',
+        'fabs', 'float32', 'float64', 'floor',
+        'int8', 'int16', 'int32', 'int64', 'isinf', 'isneginf', 'isposinf', 'isnan', 'invert', 'isfinite',
         'log', 'log10', 'log1p', 'log2', 'logical_not',
         'negative',
         'rad2deg', 'radians','reciprocal', 'rint', 'round', 'round_',
         'sign', 'signbit',
          'sin', 'sinh', 'spacing', 'sqrt', 'square',
         'tan', 'tanh','trunc',
+        'uint8', 'uint16', 'uint32', 'uint64'
         )
 
 for f in unary_ufunc:

